mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Backup.java

```diff
@@ -21,10 +21,17 @@
 import java.io.IOException;
 import java.io.OutputStream;
 import java.util.List;
+import java.util.stream.Stream;
 
+import org.apache.james.mailbox.store.mail.model.Mailbox;
 import org.apache.james.mailbox.store.mail.model.MailboxMessage;
 
 public interface Backup {
 
-    void archive(List<MailboxMessage> messages, OutputStream destination) throws IOException;
+    /**
+     * @param mailboxes list of mailboxes to be stored in the archive
+     * @param messages a stream of MailboxMessages that will be consumed
+     * @param destination an OutputStream in which the zip will be written
+     */
+    void archive(List<Mailbox> mailboxes, Stream<MailboxMessage> messages, OutputStream destination) throws IOException;
```

Comment:
yes but in this case we did not find a better way to share this big chunk of data. Do you have something better?
BTW that's why the documentation explicitly states that the stream will be consumed

Reply:
in fact the Stream abstraction is not ideal because it means that the stream magically calls a backend to load more data and the archive method has to handle failures of the backend it knows nothing about.
A much better abstraction would be to use a RxJava-like (reactive stream) : the backend could then push data in the reactive stream and have backup handling entries without sacrifying extensibility (you can have several consumer to the reactive stream).
It's probably not for today but probably something to think about for long running tasks.